src/ucp/rndv/proto_rndv.c

```diff
@@ -363,28 +346,143 @@ ucp_proto_rndv_ctrl_init(const ucp_proto_rndv_ctrl_init_params_t *params,
+    ucp_proto_init_elem_t *remote_proto;
+    ucp_proto_caps_t *remote_caps;
+    ucs_status_t status;
+    size_t min_length, max_length, cfg_thresh, cfg_priority;
+    void *remote_priv;
+
+    ucs_assert(params->super.flags & UCP_PROTO_COMMON_INIT_FLAG_RESPONSE);
+    ucs_assert(!(params->super.flags & UCP_PROTO_COMMON_INIT_FLAG_SINGLE_FRAG));
+
+    if (!ucp_proto_common_init_check_err_handling(&params->super)) {
+        status = UCS_ERR_UNSUPPORTED;
+        goto out;
+    }
+
+    /* Initialize 'rpriv' structure */
+    ucp_proto_rndv_ctrl_init_priv(params, lane);
+
+    /* Init remote proto */
+    remote_select_param = ucp_proto_rndv_remote_select_param_init(params);
+    status              = ucp_proto_rndv_ctrl_select_remote_proto(
+            params, &remote_select_param, rpriv->md_map, &remote_proto_select);
+    if (status != UCS_OK) {
+        goto out;
+    }
+    remote_proto_init = &remote_proto_select->proto_init;
+
+    /* Add variants for each remote proto */
+    ucs_array_for_each(remote_proto, &remote_proto_init->protocols) {
+        remote_caps = &remote_proto->caps;
+        /* Skip empty variants and reconfig proto */
+        if ((remote_caps->num_ranges == 0) || (remote_proto->proto_id == 0)) {
+            continue;
+        }
+
+        last_range = remote_caps->ranges + (remote_caps->num_ranges - 1);
+        max_length = ucs_min(params->super.max_length, last_range->max_length);
+        min_length = ucs_max(params->super.min_length,
+                              remote_proto->caps.min_length);
+        if (max_length < min_length) {
+            continue;
+        }
+
+        remote_priv = &ucs_array_elem(&remote_proto_init->priv_buf,
+                                      remote_proto->priv_offset);
+        ucp_proto_rndv_set_variant_config(init_params, remote_proto,
+                                          &remote_select_param, remote_priv,
+                                          &rpriv->remote_proto_config);
+
+        cfg_thresh   = params->super.cfg_thresh;
+        cfg_priority = params->super.cfg_priority;
+        if ((remote_proto->cfg_thresh != UCS_MEMUNITS_AUTO)) {
+            if (cfg_thresh == UCS_MEMUNITS_AUTO) {
+                cfg_thresh = remote_proto->cfg_thresh;
+            }
+
+            /* If some remote variant has certain thresh, we need to
+             * increase their cfg_priority to force it in case of the same
+             * params->super.cfg_thresh and remote_proto->cfg_thresh.
+             *    E.g. RNDV_THRESH=0 and RNDV_MODE=get_zcopy
+             *    In this case all the variants would have 0 cfg_thresh, but
+             *    get_zcopy should be preferred over other RNDV protocols.
+             * 
+             * +1 is required in case if remote variant cfg_priority is 0.
+             */
+            if (remote_proto->cfg_thresh != UCS_MEMUNITS_INF) {
+                cfg_priority += remote_proto->cfg_priority + 1;
+            }
```

Comment:
can we do that using thresh? If we take max of thresh values, all rndv variants will be `inf` except the one which is selected by `UCX_RNDV_SCHEME`

Reply:
No we cannot, imagine the case when `UCX_RNDV_THRESH=0` and `UCX_RNDV_SCHEME=get_zcopy` but `rndv/get/zcopy` is unavailable. If we use only `cfg_thresh` all the RNDV protocols in that case would have `inf` threshold and so the eager would be used. 

I understand that this is a kind of a corner case, but I don't think that changing current behavior is a good idea.

Reply:
Created separate a patch to simplify this logic: https://github.com/openucx/ucx/pull/9815

Reply:
The patch that was supposed to simplify that logic cannot be merged since many tests rely on that logic and some of them cannot be workarounded (see the PR mentioner above).

Reply:
is there any other drawback? Using thresholds still lokks better for me

Reply:
It was decided to try setting `cfg_priority` to the current protocol value for all variants, and initialize `cfg_thresh` as  `max(protocol_cfg_thresh, variant_cfg_thresh)`. Need to check whether it would cause many workarounds in gtest logic.